### PR TITLE
Feature/local logging

### PR DESCRIPTION
**Pull request by cweniger**
This pull request introduces a new modular logging system to the Falcon codebase, replacing the previous WandB-only approach with a flexible architecture that supports both local file-based and WandB logging backends. The configuration files are updated to use a new nested structure for logging settings, and the CLI is refactored to initialize logging based on this new configuration. The changes are backward compatible with legacy config files.

The most important changes are:

**1. Modular Logging System Implementation**
- Added a new abstract logging interface (`LoggerBackend`) and a `CompositeLogger` class to support multiple pluggable logging backends (e.g., WandB, local files) through a unified API. Includes a `LoggerManager` Ray actor for distributed logging, and a `NullBackend` for disabling logging. (`falcon/core/logger.py`)
- Introduced a local file-based logging backend (`LocalFileBackend`) and its Ray actor wrapper (`LocalLoggerActor`), along with a factory function for easy integration with the logger manager. (`falcon/core/local_logger.py`)
- Updated the Falcon core module exports to include the new logging interfaces and backends for streamlined imports. (`falcon/core/__init__.py`)

**2. CLI and Logging Initialization Refactor**
- Refactored the CLI to initialize logging using the new modular system, supporting both the new nested config structure and legacy flat structure for backward compatibility. Logging backends are now selected and configured at runtime based on the config. (`falcon/cli.py`) [[1]](diffhunk://#diff-3c2a54e715858582ae01dc5bd38fae98c73445c6d50746e4ed679afcdee60d44R23-R92) [[2]](diffhunk://#diff-3c2a54e715858582ae01dc5bd38fae98c73445c6d50746e4ed679afcdee60d44L61-R131)
- Added import for `read_run` in the Falcon init module. (`falcon/__init__.py`)

**3. Configuration File Updates**
- Updated all example configuration files (`examples/01_minimal/config.yaml`, `examples/02_bimodal/config_amortized.yaml`, `examples/02_bimodal/config_regular.yaml`, `examples/02_bimodal/config_rounds_fill.yaml`, `examples/02_bimodal/config_rounds_renew.yaml`, `examples/03_composite/config.yaml`) to use the new nested logging config structure, allowing independent enabling/disabling of WandB and local logging. (`[[1]](diffhunk://#diff-0e64ed4ff45269f625853bd0034a79dfffbe1f56b521a7f673454627d4bd3680L29-R39)`, `[[2]](diffhunk://#diff-08c8651e7232dff0bd39152583c9fc762469ce4b15db1c1ef90199fde8d7bc1cL1-R10)`, `[[3]](diffhunk://#diff-bec390d5fc2ae2d34bb257ed983d9eb0f7afc040233ed48a255f5ea23beee1bbL1-R10)`, `[[4]](diffhunk://#diff-e2f19f6af6ee3fc3687d95c1e7359eaa18f2612c567a752b939fbcd8d3d2f9d9L1-R10)`, `[[5]](diffhunk://#diff-8377e70016682fed6cc0a8fd2b824396c369ceef806407c9faf395ed586b72abL1-R10)`, `[[6]](diffhunk://#diff-8a51f8687dd610670620acee38fcd0d261cfe615c6f915c8fcb47e6801ff2ad4L1-R10)`)
- Reduced the default number of training epochs in the minimal example from 300 to 10 for faster runs. (`examples/01_minimal/config.yaml`)

### CODE REVIEW

File: examples/01_minimal/config.yaml [hunk 1]
```diff
@@ -78,7 +83,7 @@ graph:
          _input_: [x]
 
       # --- Training parameters ---
-      num_epochs: 300
```
Comment (cweniger) [inline examples/01_minimal/config.yaml]: Keep this line

File: falcon/core/logger.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            factory: Callable that creates a backend actor.
+        """
+        self.backend_factories[name] = factory
+
+    def init(self, actor_id: str, config: Optional[Dict[str, Any]] = None) -> None:
+        """Initialize all registered backends for an actor.
+
+        Args:
+            actor_id: Unique identifier for the actor.
+            config: Configuration dict passed to backend factories.
```
Comment (cweniger) [inline falcon/core/logger.py]: What is config good for?

**Comment by cweniger** [comment]
Start_wandb_logger is still starting all loggers. That is confusing sans should be refactored, so that wandb loses its apparent central role. Make sure file structures are symmetric for the local logger and wandb to reduce confusion.

File: falcon/core/wandb_logger.py [hunk 3]
```diff
@@ -1,115 +1,201 @@
...
+    }
+    if local_log_dir:
+        factories["local"] = create_local_factory(local_log_dir)
+
+    LoggerManager.options(name="falcon:global_logger", lifetime="detached").remote(
+        backend_factories=factories
+    )
+
+
+def finish_wandb_logger():
```
Comment (cweniger) [inline falcon/core/wandb_logger.py]: Move, make logger file structures clean

**Comment by cweniger** [comment]
For local logger, maybe metrics is not the optimal folder name?

**Comment by cweniger** [comment]
## Review Comments Addressed

All review comments have been addressed in the latest commits:

### 1. `examples/01_minimal/config.yaml:81` - "Keep this line"
✅ Restored `num_epochs: 300` (was accidentally changed to 10)

### 2. `falcon/__init__.py:7` - "There should be a test for this"
✅ Added comprehensive tests for `read_run` in `tests/test_run_reader.py`:
- 10 unit tests covering `MetricReader`, `NodeReader`, `RunReader`
- Tests for single/multiple chunks, array values, error handling

### 3. `falcon/cli.py:28` - "This should be part of logger.py"
✅ Moved `_init_logging` → `init_logging()` in `falcon/core/logger.py`
- CLI now imports from logger module
- Cleaner separation of concerns

### 4. `falcon/core/logger.py:160` - "What is config good for?"
✅ Removed the unused `config` parameter from `LoggerManager.init()` and all factory signatures
- Was dead code / premature abstraction
- Simplified the interface

### 5. `falcon/core/local_logger.py:55` - "Should be reasonable default"
✅ Changed `buffer_size` default from 5 → 100 with documentation explaining the trade-off:
- Lower values = more frequent saves, more I/O
- Higher values = less I/O, but risk data loss on crash
- 100 is a reasonable balance

### 6. `falcon/core/local_logger.py:14` - "Walltime should be implemented higher up?"
✅ Walltime is now generated once at the `LoggerManager.log()` level and passed to all backends
- Ensures consistent timestamps across all backends logging the same metric
- Backends accept optional `walltime` parameter (use provided or generate own)

### Additional cleanup
- Removed unused `start_wandb_logger` and `finish_wandb_logger` functions
- Added `finish_logging()` to `logger.py` as the generic shutdown function
- Added smoke tests for all 3 examples in `tests/test_examples_smoke.py`

### Tests
All 13 tests pass:
- 3 smoke tests (examples 01, 02, 03)
- 10 unit tests (run_reader)

**Review by cweniger — COMMENTED** [review]
Looks all good.
